presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryLeakDetector.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.memory;
+
+import com.facebook.presto.execution.QueryInfo;
+import com.facebook.presto.execution.QueryManager;
+import com.facebook.presto.execution.QueryState;
+import com.facebook.presto.spi.QueryId;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import io.airlift.log.Logger;
+import org.joda.time.DateTime;
+import org.weakref.jmx.Managed;
+
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.facebook.presto.execution.QueryState.RUNNING;
+import static com.facebook.presto.memory.LocalMemoryManager.GENERAL_POOL;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.joda.time.Seconds.secondsBetween;
+
+public class ClusterMemoryLeakDetector
+{
+    private static final Logger log = Logger.get(ClusterMemoryLeakDetector.class);
+
+    // It may take some time to remove a query's memory reservations from the worker nodes,
+    // that's why we check to see whether some time has passed after the query end time to claim that
+    // a query is leaked.
+    private static final long DEFAULT_LEAK_CLAIM_DELTA_SECONDS = 60;
+
+    private final ClusterMemoryManager clusterMemoryManager;
+    private final QueryManager queryManager;
+    private final AtomicInteger numberOfLeakedQueries = new AtomicInteger();
+    private final ScheduledExecutorService leakDetectorScheduler = newSingleThreadScheduledExecutor(daemonThreadsNamed("cluster-leak-detector"));
+
+    @Inject
+    public ClusterMemoryLeakDetector(ClusterMemoryManager clusterMemoryManager, QueryManager queryManager)
+    {
+        this.clusterMemoryManager = requireNonNull(clusterMemoryManager, "clusterMemoryManager is null");
+        this.queryManager = requireNonNull(queryManager, "queryManager is null");
+    }
+
+    @PostConstruct
+    public void start()
```

Comment:
add `@PreDestroy` method shutting down `leakDetectorScheduler` and use that in tests.

Reply:
What do you mean by "use that in tests"? When the test Presto server is shutdown it will stop the lifecycle and call this new shutdown method.